beacon-chain/rpc/eth/beacon/pool.go

```diff
@@ -306,8 +307,61 @@ func (bs *Server) SubmitVoluntaryExit(ctx context.Context, req *ethpbv1.SignedVo
 	return &emptypb.Empty{}, nil
 }
 
+// SubmitSignedBLSToExecutionChanges submits said object to the node's pool
+// if it passes validation the node must broadcast it to the network.
+func (bs *Server) SubmitSignedBLSToExecutionChanges(ctx context.Context, req *ethpbv2.SubmitBLSToExecutionChangesRequest) (*emptypb.Empty, error) {
+	ctx, span := trace.StartSpan(ctx, "beacon.SubmitVoluntaryExit")
+	defer span.End()
+	st, err := bs.ChainInfoFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not get head state: %v", err)
+	}
+	var failures []*helpers.SingleIndexedVerificationFailure
+	for i, change := range req.GetChanges() {
+		alphaChange := migration.V2SignedBLSToExecutionChangeToV1Alpha1(change)
+		_, err = blocks.ValidateBLSToExecutionChange(st, alphaChange)
+		if err != nil {
+			failures = append(failures, &helpers.SingleIndexedVerificationFailure{
+				Index:   i,
+				Message: "Could not validate SignedBLSToExecutionChange: " + err.Error(),
+			})
+			continue
+		}
+		if err := blocks.VerifyBLSChangeSignature(st, change); err != nil {
+			failures = append(failures, &helpers.SingleIndexedVerificationFailure{
+				Index:   i,
+				Message: "Could not validate signature: " + err.Error(),
+			})
+			continue
+		}
+		bs.BLSChangesPool.InsertBLSToExecChange(alphaChange)
+		if st.Version() >= version.Capella {
+			if err := bs.Broadcaster.Broadcast(ctx, alphaChange); err != nil {
+				failures = append(failures, &helpers.SingleIndexedVerificationFailure{
+					Index:   i,
+					Message: "Could not broadcast BLSToExecutionChange: " + err.Error(),
+				})
+				continue
+			}
+		}
+	}
+	if len(failures) > 0 {
+		failuresContainer := &helpers.IndexedVerificationFailure{Failures: failures}
+		err := grpc.AppendCustomErrorHeader(ctx, failuresContainer)
+		if err != nil {
+			return nil, status.Errorf(
+				codes.InvalidArgument,
+				"One or more BLSToExecutionChange failed validation. Could not prepare BLSToExecutionChange failure information: %v",
+				err,
+			)
+		}
+		return nil, status.Errorf(codes.InvalidArgument, "One or more BLSToExecutionChange failed validation")
```

Comment:
To be most informative, why don't we tell the user that all other indices not in the failure list succeeded and were included in the pool + broadcast to the network? People might think a failure could mean none of them succeeded. Can we included the indices in the error?

Reply:
The nicest UX would be to tell the user the validator indices for which the failures correspond to

Reply:
Dito above, James' PR is the user facing end. This implements the backend only

Reply:
👍